### ISSUE TITLE
Rule proposal: `prefer-array-find`

### ISSUE THREAD

**Issue by oscarmarcelo** [sindresorhus/eslint-plugin-unicorn#730]
Prefer `.find()` over `.filter()[0]`, since `find` breaks the loop as soon as it finds a match.

## Fail
```js
const item = array.filter(x => x === '🦄')[0];
```

## Pass
```js
const item = array.find(x => x === '🦄');
```

**Comment by fisker** [comment]
# Fail

```js
const item = array.filter(x => x === '🦄').unshift();
```

**Comment by sindresorhus** [comment]
Accepted. PR welcome.

**Comment by fisker** [comment]
sindresorhus Asked about this in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/735#discussion_r426395594

```js
const result = array.filter(x => x === '🦄');
const item = result[0];
```


To support this, we need check `result` not exported, and only accessed with `[0]`


1. If we fix to 

```js
const result = array.find(x => x === '🦄');
const item = result;
```

The name may confuse people `const allTrueArray = array.filter(foo)` -> `const allTrueArray = array.find(foo)` the `allTrueArray` is not an array anymore. 

2. If we fix to

```js
const item = array.find(x => x === '🦄');
```

problem is the `item` may in a loop, this will call `.find()` may times.

3. If we support this, should we support

```js
const result = array.filter(x => x === '🦄');
const item = result.shift();
```

```js
const result = array.filter(x => x === '🦄');
const [item] = result;
```

too?


@oscarmarcelo  What do you think?

**Comment by oscarmarcelo** [comment]
Honestly, I don't understand the implications of the first two points, but I totally agree with the third one!

**Comment by fisker** [comment]
I have one concern about side effects of `.filter` call, in this case

```js
const arrayTypeObjects = foo.filter(object => {
	const isArray = Array.isArray(object.data);
	object.isArray = isArray; // <-- also add a mark to every object
	return isArray;
});

const firstArray = arrayTypeObjects[0];

doSomething(firstArray);

for (const object of foo) {
	if (object.isArray) { // <-- rely on the mark set above
		doSomething(object.data)
	} else {
		doSomethingDifferent(object.data)
	}
}
```

If we fix `foo.filter` to `foo.find`, the code will be broken.

For this reason, I suggest auto-fix under an option like this

```js
{
	'unicorn/prefer-array-find': [
		'error',
		{
			ignoreSideEffect: true, 
			operatorForDefaultValueOfDestructuring: '??'
		}
	
}
```

### `ignoreSideEffect`

Default `false`, set to `true` enables auto fix.

### `operatorForDefaultValueOfDestructuring`

Default `undefined`, 

For this case

```js
const [foo = bar] = array.filter(fn);
```

#### `undefined`

Suggest `const foo = array.filter(fn) ?? bar` and `const foo = array.filter(fn) || bar`, no auto fix.

#### `nullish-coalescing-operator` / `??` 

Auto fix to  `const foo = array.filter(fn) ?? bar`

#### `logical-or-operator` / `||` 

Auto fix to  `const foo = array.filter(fn) || bar`


//cc @sindresorhus @oscarmarcelo Thoughts?

**Comment by oscarmarcelo** [comment]
Nice find! I didn't think about this to this level of detail! 😃
LGTM! 👍

**Comment by fisker** [comment]
I thought this rule is easy too, actually it only took about 30 mins to finish the first implementation, then I dig more and more deeper...
